taichi/analysis/same_statements.cpp

```diff
@@ -219,8 +277,12 @@ class IRNodeComparator : public IRVisitor {
   static bool run(IRNode *root1,
                   IRNode *root2,
                   const std::optional<std::unordered_map<int, int>> &id_map,
-                  bool check_same_value) {
-    IRNodeComparator comparator(root2, id_map, check_same_value);
+                  bool check_same_value,
+                  const std::optional<std::unordered_set<AsyncState>>
```

Comment:
nit: could you add a comment explaining why we need to use `std::optional<std::unordered_*>`, instead of just an `std::unordered_*` (with empty being equivalent to `std::nullopt`)?

Reply:
I need to distinguish the case of an empty unordered_set (assuming every SNodes are unchanged) and empty (assuming nothing). Maybe it's better to pass in another boolean variable?